zipextimporter.py

```diff
@@ -42,138 +43,227 @@
-                return self
-        return None
-
-    def find_spec(self, name, path=None):
-        module = self.find_module(name, path)
-        if module is not None:
-            return util.spec_from_loader(name, module)
+
+class _ModuleInfo:
+    __slots__ = ("path", "is_ext", "is_package")
+    def __init__(self, *args):
+        self.path, self.is_ext, self.is_package = args
+
+
+class ZipExtensionImporter(zipimporter):
+    suffixes = _imp.extension_suffixes()
+    _searchorder = (
+        *[(f"\\__init__{suffix}", True, True) for suffix in suffixes],
+        ("\\__init__.pyc", False, True),
+        ("\\__init__.py", False, True),
+        *[(suffix, True, False) for suffix in suffixes],
+        (".pyc", False, False),
+        (".py", False, False),
+    )
+    _searchorder_pywin32 = (
+        *_searchorder[:-2],
+        ("%d%d.dll" % sys.version_info[:2], True, False),
+        *_searchorder[-2:],
+    )
+    del suffixes
+    names_pywin32 = frozenset(["pywintypes", "pythoncom"])
+    verbose = _memimporter.get_verbose_flag()
+
+    def __init__(self, path_or_importer):
+        if isinstance(path_or_importer, zipimporter):
+            self.zipimporter = path_or_importer
         else:
-            return None
-
-    def load_module(self, fullname):
-        verbose = _memimporter.get_verbose_flag()
-
-        if fullname in sys.modules:
-            mod = sys.modules[fullname]
-            if verbose:
-                sys.stderr.write(
-                    "import %s # previously loaded from zipfile %s\n"
-                    % (fullname, self.archive))
-            return mod
+            self.zipimporter = zipimporter(path_or_importer)
+            if hasattr(zipimporter, "_files"):  # py <= 37, built-in
+                # will never enter here, because droped py37
+                super().__init__(path_or_importer)
+
+    def __getattr__(self, name):
+        return getattr(self.zipimporter, name)
+
+    @property
+    def zipextimporter(self):
+        if isinstance(self, ZipExtensionImporter):
+            return self
         try:
-            return zipimport.zipimporter.load_module(self, fullname)
-        except ImportError as err:
-            if verbose:
-                sys.stderr.write("error loading %s: %s\n"% (fullname, err))
-
-            if sys.version_info >= (3, 0):
-                # name of initfunction
-                initname = "PyInit_" + fullname.split(".")[-1]
-            else:
-                # name of initfunction
-                initname = "init" + fullname.split(".")[-1]
-            filename = fullname.replace(".", "\\")
-            if filename in ("pywintypes", "pythoncom"):
-                filename = filename + "%d%d" % sys.version_info[:2]
-                suffixes = ('.dll',)
+            zipextimporter = self._zipextimporter
+        except AttributeError:
+            self._zipextimporter = zipextimporter = ZipExtensionImporter(self)
+        return zipextimporter
```

Comment:
All this (plus the `__init__` manipulation) feels like a hack. Could you explain why it was necessary? `zipextimporter` was just a regular child class of `zipimporter`, and I would appreciate to keep it that way... Do you think it would be possible to restore this status? 

In general, I like the new methods provided by this MR, but I do not see why we need to overhaul the class base structure to implement them.

Reply:
All this are needed by monkey patch. As I said above, I have made changes, only for improve compatibility. Please imagine that, if someone wrote some code about `zipimporter`, he can run it with `zipextimporter` without any modification.

Reply:
In the current implementation, if someone wants to extend `zipimporter` they would most likely write a wrapper class around it as we did, inheriting from it. If that someone then wants to extend `zipextimporter`, instead, they would just have to change their parent class.

Conversely, in the implementation proposed in this PR, extending `zipextimporter` would be impossible. 

I appreciate the additions to `zipextimporter` provided in this PR, however I am strongly against abandoning the status of "simple" child of `zipimport`. If you could accommodate your changes without interfering with the child status of `zipextimporter` (no `hasattr` magic, to be more clear) or with the original code of `zipimporter`, that would be great. Otherwise, I am afraid I would have to reject this PR and (if you are not against this) I will  just take the changes that are not disruptive manually from your repository...

Reply:
In fact extending `zipextimporter` is possible, but monkey patch is useless in this case, we need install the child to `path_hooks` just like current which the way is always available.

I know the logic complexity has increased, but there has only made addition and no replacement, all old usages were kept. If you agree with this, I will go on to work on issues which you pointed out above.